### ISSUE TITLE
How to perform action in HttpHandler both before and after invoking next

### ISSUE THREAD

**Issue by cmeeren** [giraffe-fsharp/Giraffe#326]
I have an HttpHandler like this:

```f#
let useSomething : HttpHandler =
  fun next ctx ->
    use __ = somethingToBeDisposedAtTheEndOfTheRequest
    next ctx
```

It's in my routing like this:

```f#
subRoute "..."
>=> useSomething
>=> choose [
  routex "..." >=> GET >=> myGet
  routex "..." >=> POST >=> myPost
  ...
]
```

However, the disposable seems to always be disposed before `myGet` or `myPost` is called. Have I misunderstood something regarding the middleware pipeline metaphor? How can I ensure that the disposable is disposed only after e.g. `myGet` and `myPost` have executed?

**Comment by cmeeren** [comment]
Sorry, I'm still a bit lost. I tried

```
let useSomething handler : HttpHandler =
  fun next ctx ->
    use __ = somethingToBeDisposedAtTheEndOfTheRequest
    handler next ctx
```

and

```f#
subRoute "..."
>=> useSomething (choose [
  routex "..." >=> GET >=> myGet
  routex "..." >=> POST >=> myPost
  ...
])
```

but I'm still observing the same behavior. What am I doing wrong?

**Comment by slang25** [comment]
You might need to add `task` and `let!` to avoid the extra bit of deferred execution (I think)

**Comment by slang25** [comment]
```fsharp
let useSomething handler : HttpHandler =
  fun next ctx -> task {
    use __ = somethingToBeDisposedAtTheEndOfTheRequest
    do! handler next ctx
  }
```

**Comment by cmeeren** [comment]
Yes, that did the trick. Thanks!
